shared/lsif/providers.ts

```diff
@@ -30,55 +39,289 @@ export function createProviders(logger: Logger): Providers {
-        yield* (await provider).hover(doc, pos)
-    }
+        position: sourcegraph.Position
+    ): Promise<sourcegraph.Hover | null> => {
+        const query = `
+            query Hover($repository: String!, $commit: String!, $path: String!, $line: Int!, $character: Int!) {
+                repository(name: $repository) {
+                    commit(rev: $commit) {
+                        blob(path: $path) {
+                            lsif {
+                                hover(line: $line, character: $character) {
+                                    markdown {
+                                        text
+                                    }
+                                    range {
+                                        start {
+                                            line
+                                            character
+                                        }
+                                        end {
+                                            line
+                                            character
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        `
 
-    return {
-        definition,
-        references,
-        hover,
+        const lsifObj: HoverResponse | null = await queryLSIF(
+            {
+                doc,
+                position,
+                query,
+            },
+            queryGraphQL
+        )
+
+        if (!lsifObj || !lsifObj.hover) {
+            return null
+        }
+
+        return {
+            contents: {
+                value: lsifObj.hover.markdown.text,
+                kind: sourcegraph.MarkupKind.Markdown,
+            },
+            range: lsifObj.hover.range,
+        }
     }
 }
 
 /**
- * Return the GraphQL LSIF providers if the Sourcegraph instance supports it.
- * Otherwise, use the HTTP API providers.
+ * Perform an LSIF request to the GraphQL API.
+ *
+ * @param args Parameter bag.
+ * @param queryGraphQL The function used to query the GraphQL API.
  */
-async function selectProvider(logger: Logger): Promise<Providers> {
-    const supportsGraphQL = compareVersion({
-        productVersion: await productVersion(),
-        minimumVersion: GRAPHQL_API_MINIMUM_VERSION,
-        minimumDate: GRAPHQL_API_MINIMUM_DATE,
-    })
-
-    if (supportsGraphQL) {
-        logger.log('Sourcegraph instance supports LSIF GraphQL API')
-        return createGraphQLProviders()
+async function queryLSIF<
+    P extends {
+        doc: sourcegraph.TextDocument
+        position: lsp.Position
+        query: string
+    },
+    R
+>(
+    {
+        /** The current text document. */
+        doc,
+        /** The current hover position. */
+        position,
+        /** The GraphQL request query. */
+        query,
+        /** Additional query parameters. */
+        ...rest
```

Comment:
I guess this is probably the existing behaviour, but should we consider doing this spread first so the trailing args can't override the essential keys?

Reply:
So the spread operator MUST be the last for params, but not for object construction (which we do later in this method). It's a good point, updating.